src/display/annotation_layer.js

```
@@ -38,341 +38,531 @@ var CustomStyle = displayDOMUtils.CustomStyle;
-        // to adjust for all borders. This yields a similar result as
-        // Adobe Acrobat/Reader.
-        width = width - 2 * data.borderStyle.width;
-        height = height - 2 * data.borderStyle.width;
-      }
+/**
+ * @typedef {Object} AnnotationElementParameters
+ * @property {Object} data
+ * @property {PDFPage} page
+ * @property {PageViewport} viewport
+ * @property {IPDFLinkService} linkService
+ */
 
-      var horizontalRadius = data.borderStyle.horizontalCornerRadius;
-      var verticalRadius = data.borderStyle.verticalCornerRadius;
-      if (horizontalRadius > 0 || verticalRadius > 0) {
-        var radius = horizontalRadius + 'px / ' + verticalRadius + 'px';
-        CustomStyle.setProp('borderRadius', container, radius);
-      }
+/**
+ * @class
+ * @alias AnnotationElementFactory
+ */
+function AnnotationElementFactory() {}
+AnnotationElementFactory.prototype =
+    /** @lends AnnotationElementFactory.prototype */ {
+  /**
+   * @param {AnnotationElementParameters} parameters
+   * @returns {AnnotationElement}
+   */
+  create: function AnnotationElementFactory_create(parameters) {
+    var subtype = parameters.data.annotationType;
+
+    switch (subtype) {
+      case AnnotationType.LINK:
+        return new LinkAnnotationElement(parameters);
 
-      switch (data.borderStyle.style) {
-        case AnnotationBorderStyleType.SOLID:
-          container.style.borderStyle = 'solid';
-          break;
+      case AnnotationType.TEXT:
+        return new TextAnnotationElement(parameters);
 
-        case AnnotationBorderStyleType.DASHED:
-          container.style.borderStyle = 'dashed';
-          break;
+      case AnnotationType.WIDGET:
+        return new WidgetAnnotationElement(parameters);
 
-        case AnnotationBorderStyleType.BEVELED:
-          warn('Unimplemented border style: beveled');
-          break;
+      default:
+        throw new Error('Unimplemented annotation type "' + subtype + '"');
```

What is the reason to throw error instead of warn it and just don't render anything? I don't think there are any code that catches this error so this error is a fatal error that will propagate to top level and stop the annotation rendering process. I think this is also the reason why #6818 failed (but I still prefer #6786 why of doing it)